internal/host/hostcommands/instruction_manager.go

```diff
@@ -178,10 +190,90 @@ func (i *InstructionManager) GetNextSteps(ctx context.Context, host *models.Host
 	} else {
 		returnSteps.NextInstructionSeconds = defaultNextInstructionInSec
 	}
+
+	// If the agent isn't compatible with the service and the host is in a state that allows to
+	// upgrade the agent, then replace all the calculated steps with a single step to upgrade
+	// the agent:
+	if i.isAgentUpgradeAllowed(host) && !i.isAgentCompatible(host) {
+		log.WithFields(logrus.Fields{
+			"expected_image": i.config.AgentImage,
+			"actual_image":   host.DiscoveryAgentVersion,
+		}).Info(
+			"Agent image ins't compatible with the service, and it can be upgraded, " +
+				"will replace all the calculated steps with a single step to " +
+				"upgrade the image",
+		)
+		var err error
+		returnSteps.Instructions, err = i.upgradeAgentCmd.GetSteps(ctx, host)
+		if err != nil {
+			return models.Steps{}, err
+		}
+		returnSteps.PostStepAction = swag.String(models.StepsPostStepActionContinue)
+		returnSteps.NextInstructionSeconds = defaultNextInstructionInSec
+		events.SendUpgradeAgentStartedEvent(
+			ctx,
+			i.eventsHandler,
+			*host.ID,
+			hostutil.GetHostnameForMsg(host),
+			host.InfraEnvID,
+			host.ClusterID,
+			i.config.AgentImage,
+		)
+	}
+
 	logSteps(returnSteps, InfraEnvID, hostID, log)
 	return returnSteps, nil
 }
 
+// isAgentCompatible checks if the agent running in the host is compatible with what the service
+// expects.
+func (i *InstructionManager) isAgentCompatible(host *models.Host) bool {
```

Comment:
why we need this wrapper function?

Reply:
The function in the common package is needed because it is also used by the validation. This one isn't really needed, just convenient to simplify a bit the `if` where it is called. And it is also a good place for comments. I can remove it if you want.

Reply:
i prefer to use same function in common

Reply:
Done.